---
title: "Sub_0645: miRNA Normalization "
subtitle: "Sensitivity to PCR bias and dilution"
author: "John Steill"
date: "`r Sys.Date()`"
output:
  tufte::tufte_handout:
    citation_package: natbib
    latex_engine: xelatex
  tufte::tufte_book:
    citation_package: natbib
    latex_engine: xelatex
  tufte::tufte_html: default
bibliography: skeleton.bib
link-citations: yes
---
```{r global_options, include=FALSE}
knitr::opts_chunk$set(fig.pos = 'h')
```
```{r setup, include=FALSE}
library(tufte)
library(tidyverse)
library(reshape2)
library(edgeR)
library(xtable)
# invalidate cache when the tufte version changes
knitr::opts_chunk$set(tidy = FALSE, cache.extra = packageVersion('tufte'))
options(htmltools.dir.version = FALSE)
setwd("/Users/jsteill/GDrive/PROJS/bmi776/Spring-18/project")
```

# Introduction

There are two goals for this analysis:

1. To evaluate the effect of rna concentration and pcr on expression estimation 
1. Create normalization method:
    + C. Elegans spike-in
    + CPM vs tam_UQ 

Unfortunately, the spike-in method failed. No reads were produced. 

In evaluating the noise induced by lower concentrations and pcr amplification, we are mostly concerned with the abundances of the most prevalent miRNAs. We are interested in when the values are distorted to the extent that their ranks change. For this question, choice of normalization method won't matter: whether we divide by sum()/1e6 (cpm) or by a given quantile value, the ranks will be unaffected.

However, this data should provide insights into the best normalization method. As the base sample is the same for each trial, all differentially expressed miRNAs will be false positives. Thus we may want to minimize the FDR by choosing the method with the fewest DE miRNAs.

\pagebreak

# Raw Counts

```{r, include=FALSE}


get_data <- function(f){
    raw <- read_tsv(f)
    raw = raw %>% dplyr::rename(
        miRNA="#symbol",
        ng_100_gold = 'Hepg2_with_UO126_reps_1&2_100ng')
    raw = raw[,c(1,2,3,12,4,13,5,14,6,15,7,16,8,17,9,18,10,19,11,20)]
    raw$rowSum = rowSums(raw[,-1])
    raw <- arrange(raw,desc(rowSum))[,1:ncol(raw)-1]
}


get_melt <- function(f){
    m = melt(get_data(f)[1:10,])  
}

```

```{r fig-raw, fig.width = 10, fig.height = 6, fig.fullwidth = TRUE, fig.cap = "Raw Reads", message = F, warning=FALSE, echo =F, cache=TRUE}
setwd("/Users/jsteill/GDrive/PROJS/bmi776/Spring-18/project")

p <- ggplot(get_melt("Data/miRNA.ec.tab"), aes(x=variable, y=value, group=miRNA, color=miRNA))+ geom_line(aes(linetype=miRNA)) + coord_trans(y = "log10") + theme(axis.text.x=element_text(angle=45, hjust=1)) + labs(y = "Raw counts", x = '')

p2 <- ggplot(get_melt("Data/cpm_norm_sub_670.tsv"), aes(x=variable, y=value, group=miRNA, color=miRNA))+ geom_line(aes(linetype=miRNA)) + coord_trans(y = "log10") + theme(axis.text.x=element_text(angle=45, hjust=1)) + labs(y = "CPM", x = '')

p3 <- ggplot(get_melt("Data/tam_UQ_sub_670.tsv"), aes(x=variable, y=value, group=miRNA, color=miRNA))+ geom_line(aes(linetype=miRNA)) + coord_trans(y = "log10") + theme(axis.text.x=element_text(angle=45, hjust=1)) + labs(y = "Tam_UQ", x = '')

raw = get_data("Data/miRNA.ec.tab")
## Following taken from Differential Expression Analysis using edgeR - 
## https://bioinformatics-core-shared-training.github.io/cruk-bioinf-sschool/Day3/rnaSeq_DE.pdf
dgList <- DGEList(counts=raw[,-1], genes=raw[,1])
countCheck <- countsPerMillion > 1
keep <- which(rowSums(countCheck) >= 2)
dgList <- dgList[keep,]
dgList <- calcNormFactors(dgList, method="upperquartile", P=.999999)
plot_data = melt(cbind(dgList$genes, dgList$counts)[1:10,])
p4 <- ggplot(plot_data, aes(x=variable, y=value, group=miRNA, color=miRNA))+ geom_line(aes(linetype=miRNA)) + coord_trans(y = "log10") + theme(axis.text.x=element_text(angle=45, hjust=1)) + labs(y = "TMM", x = '')


sampleCycle = append(c("C15"),paste("C", rep(c("15","20","30"), each = 2, 3), sep="" ))
sampleType<- rep("N", ncol(dgList))

sampleType[grep("3", colnames(dgList))] <- "3"
sampleType[grep("6", colnames(dgList))] <- "6"
sampleType[grep("10", colnames(dgList))] <- "10"
sampleType[grep("gold", colnames(dgList))] <- "G"
sampleType <- factor(sampleType)
sampleType <- relevel(sampleType,"G")
sampleCycle <- factor(sampleCycle)
sampleCycle <- relevel(sampleCycle,"C15")
designMat <- model.matrix(~sampleType + sampleCycle)
estimateGLMCommonDisp(dgList, design=designMat)
dgList <- estimateGLMCommonDisp(dgList, design=designMat)
dgList <- estimateGLMTrendedDisp(dgList, design=designMat)
dgList <- estimateGLMTagwiseDisp(dgList, design=designMat)
plotBCV(dgList)
fit <- glmFit(dgList, designMat)
print(xtable(cbind(fit$genes,fit$unshrunk.coefficients )[1:12,], type = "latex"))

edgeR_result <- topTags(lrt)
edgeR_result
deGenes <- decideTestsDGE(lrt, p=0.001)
deGenes <- rownames(lrt)[as.logical(deGenes)]
plotSmear(lrt, de.tags=deGenes)
abline(h=c(-1, 1), col=2)
 print(length(deGenes))
```

We see some trends: 

* Fewer total counts as the initial concentration drops 
* Fewer total counts with lower cycle counts (saturating at ~15) 
* Robust total counts in the dilution sequence, although the 1:15.6 appears to be an outlier.

\pagebreak

# CPM Normalized Counts

```{r fig-cpm, fig.width = 10, fig.height = 6, fig.fullwidth = TRUE, fig.cap = "CPM", message = F, warning=FALSE, echo =F, cache=TRUE}
setwd("/Volumes/home/GUP/Analysis/Sub_0645_miRNA_Norm")
ggplot(get_melt("data/cpm_norm.tsv"), aes(x=variable, y=value, group=miRNA, color=miRNA))+ geom_line(aes(linetype=miRNA)) + coord_trans(y = "log10") + theme(axis.text.x=element_text(angle=45, hjust=1))
```

What I find interesting here is that the relative rank of two species, 148a-3p & 151a-3p vary quite a bit in the range of supposed stability: 100-500ng and dilutions of 7.8 through 31.2. Are there two outliers?

\pagebreak

# Tam_UQ Normalized Counts
```{r fig-tam, fig.width = 10, fig.height = 6, fig.fullwidth = TRUE, fig.cap = "Tam UQ", message = F, warning=FALSE, echo =F, cache=TRUE}
setwd("/Volumes/home/GUP/Analysis/Sub_0645_miRNA_Norm")
ggplot(get_melt("data/tam_UQ.tsv"), aes(x=variable, y=value, group=miRNA, color=miRNA))+ geom_line(aes(linetype=miRNA)) + coord_trans(y = "log10") + theme(axis.text.x=element_text(angle=45, hjust=1))
```

The function of upper quantile normalization is shown by the smoothing of the highest counts. 
\pagebreak

# Required Amount Analysis: Spearman Correlation

It satisfies common sense that the very highest expressed miRNAs will be relatively unaffected by initial concentration.By doing a spearman correlation, we can examine the sensitivity of species which are express at lower levels. 

<!-- #```{r fig-spear, fig.width = 10, fig.height = 6, fig.fullwidth = TRUE,  message = F, warning=FALSE, echo =F, -->

```{r fig-spear, fig.cap = "Tam UQ Spearman Correlation; 400 top miRNA", fig.width=6, fig.height=6,  message = F, cache=TRUE, warning=FALSE, echo=F}
setwd("/Volumes/home/GUP/Analysis/Sub_0645_miRNA_Norm")
library(corrplot)
uq_df_ng = select(get_data("data/tam_UQ.tsv"),contains("ng"))
corrplot.mixed(cor(uq_df_ng[0:400,],method='spearman'), tl.pos = 'lt', lower.col = "black", number.cex = .9, order="hclust")
```

<!-- This very flat distribution is caused by the many zeros of the DF. By restricting to the top 400 expressed miRNAs: -->
<!-- ```{r fig-spear4, fig.width = 10, fig.height = 6, fig.fullwidth = TRUE, fig.cap = "Tam UQ Spearman Correlation; 400 miRNA", message = F, warning=FALSE, echo =F, cache=TRUE} -->
<!-- setwd("/Volumes/home/GUP/Analysis/Sub_0645_miRNA_Norm") -->

<!-- uq_df_ng = select(get_data("data/tam_UQ.tsv"),contains("ng")) -->
<!-- corrplot.mixed(cor(uq_df_ng[0:400,],method='spearman'), lower.col = "black", number.cex = .7, order="hclust") -->
<!-- ``` -->


# TO-DO

1. Normalization Analysis
    + Choose a region of interest (avoid outlying conditions, omit outliers?)
    + Choose a DE algorithm
    + Investigate which normalization gives the minimal DE set.
1. Revisit Spike-in's which failed.
1. ?